### PR TITLE
OpenVINO extensions

### PR DESCRIPTION
**Pull request by dkurt**
1. Convert to ONNX
```
python to_onnx.py -m models/raft-chairs.pth -o model.onnx
```

2. Convert to OpenVINO IR (use OpenVINO master branch or release >= 2021.3 because 2021.2 got stuck)
```
python mo.py --input_model model.onnx --extension extensions/mo_extension/
```

3. Build CPU extensions
```
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Release extensions/cpu_extensions
make -j$(nproc --all)
```

4. Benchmark
```
python /opt/intel/openvino_2021/deployment_tools/tools/benchmark_tool/benchmark_app.py -m model.xml -l extensions/cpu_extensions/build/libuser_cpu_extension.so
```

### CODE REVIEW

**Comment by ngaloppo** [comment]
Fantastic work @dkurt. I was able to reproduce your instructions. I'll think about validation strategies tomorrow, and also measure on TGL.

**Comment by ngaloppo** [comment]
TGL benchmark measurements: 

| Device | API | Throughput / Latency |
| ------- | --- | ---------------------- |
| CPU     | async | 0.42 FPS |

I tried `HETERO:GPU,CPU`, but unfortunately I ran into the following error when loading the model: 
```
RuntimeError: Node 1248/Unsqueeze was not assigned on any pointed device.
```

Next up I'll study GPU custom layers, hopefully pretty straight forward given the CPU extension sample code you provided.

**Comment by ngaloppo** [comment]
@dkurt when loading the RAFT network with the GPU extensions from [here](https://github.com/dkurt/openvino_pytorch_layers/pull/16), I get the following error message: 

```
[Step 7/11] Loading the model to the device
[ ERROR ] Invalid dimensions size(7) for clDNN tensor
Traceback (most recent call last):
  File "/home/ncgalopp/intel/openvino_2021.3.338/python/python3.7/openvino/tools/benchmark/main.py", line 82, in run
    if args.path_to_extension:
  File "/home/ncgalopp/intel/openvino_2021.3.338/python/python3.7/openvino/tools/benchmark/benchmark.py", line 75, in load_network
    num_requests=1 if self.api_type == 'sync' else self.nireq or 0)
  File "ie_api.pyx", line 336, in openvino.inference_engine.ie_api.IECore.load_network
  File "ie_api.pyx", line 345, in openvino.inference_engine.ie_api.IECore.load_network
RuntimeError: Invalid dimensions size(7) for clDNN tensor
```

Have you seen this before? I can't find any tensor in the IR that is more than 4 dimensions...

**Comment by dkurt** [comment]
@ngaloppo, can you share GPU extensions code by PR?

**Comment by ngaloppo** [comment]
> @ngaloppo, can you share GPU extensions code by PR?

https://github.com/dkurt/openvino_pytorch_layers/pull/16

**Comment by ngaloppo** [comment]
Note there is a bug (?) that requires CPU extensions to be loaded in addition to clDNN custom op, otherwise IE complains "opset 'extension'" not supported. 

benchmark_app only loads extensions for the target device, I had to force it to load both.

**Comment by dkurt** [comment]
@ngaloppo, there are layers with 7D dimensions op:
```xml
		<layer id="4925" name="873" type="Reshape" version="opset1">
			<data special_zero="true"/>
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>2</dim>
					<dim>3</dim>
					<dim>3</dim>
					<dim>55</dim>
					<dim>128</dim>
				</port>
				<port id="1">
					<dim>7</dim>
				</port>
			</input>
			<output>
				<port id="2" names="873" precision="FP32">
					<dim>1</dim>
					<dim>2</dim>
					<dim>9</dim>
					<dim>1</dim>
					<dim>1</dim>
					<dim>55</dim>
					<dim>128</dim>
				</port>
			</output>
		</layer>
```

**Comment by ngaloppo** [comment]
@dkurt I looked into adding 7th dimension support in clDNN. Unfortunately it may be more complicated than hoped for. Looking at the model topology, I'm wondering if we can avoid so many dimension by clever reshaping/merging nodes, since multiple dimensions are of size 1. Thoughts?

**Comment by dkurt** [comment]
@ngaloppo, I agree with you. Yes, there is an option to reduce batch dimension this way.

File: core/raft.py [hunk 1]
```diff
@@ -86,6 +86,9 @@ def upsample_flow(self, flow, mask):
     def forward(self, image1, image2, iters=12, flow_init=None, upsample=True, test_mode=False):
         """ Estimate optical flow between pair of frames """
 
+        if not iters:
+            iters = self.args.iters
```
Comment (dkurt) [inline core/raft.py:90]: `iters` has default value 12 so it won't work

**Pull request closed without merging**